### PR TITLE
Improvements to out_of_band hook

### PR DESCRIPTION
**Pull request by wjordan**
### Description
Followup to #2230 (extending the idea to move the OOB hook from `Server` into `ThreadPool`), addressing the two extra edge cases described at https://github.com/puma/puma/pull/2230#issuecomment-618825848.

Also:
- Refactor existing out-of-band tests from the big `TestPumaServer` class into its own `TestOutOfBandServer` class
- Add extra tests covering the new edge-cases handled
- A bit more cleanup/documentation on the test suite (added high-level comments describing the behavior expected by each test)

(Turns out there are a ton of complicated little edge-cases to consider around what seemed like such a simple feature!)

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` the pull request title.
- [x] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [ ] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed, including Rubocop.

### CODE REVIEW

File: lib/puma/server.rb [hunk 1]
```diff
@@ -326,6 +317,20 @@ def handle_servers
...
+    rescue Errno::ECONNABORTED
+      # client closed the socket even before accept
+      begin
+        sock.close
+      rescue
+        Thread.current.purge_interrupt_queue if Thread.current.respond_to? :purge_interrupt_queue
+      end
+    end
+    private :accept_socket
+
```
Comment (wjordan) [inline lib/puma/server.rb:333]: Extracted this logic into a private method both to simplify the accept-loop, but more importantly to allow the to be stubbed in the `test_blocks_new_connection` unit-test.

File: lib/puma/server.rb [hunk 2]
```diff
@@ -278,27 +281,15 @@ def handle_servers
...
-                  end
-                rescue SystemCallError
-                  # nothing
-                rescue Errno::ECONNABORTED
-                  # client closed the socket even before accept
-                  begin
-                    io.close
-                  rescue
-                    Thread.current.purge_interrupt_queue if Thread.current.respond_to? :purge_interrupt_queue
+                pool.wait_until_not_full
```
Comment (wjordan) [inline lib/puma/server.rb:284]: Moving the call to `wait_until_not_full` before `accept_socket` prevents connections from being accepted until out_of_band hooks are completed.

(I'm not 100% sure this change won't have any other side-effects, but the test suite seems to think it's okay...)

An alternative would be to just add an extra `pool.with_mutex {}` line above `accept_socket`.

File: lib/puma/server.rb [hunk 3]
```diff
@@ -224,8 +224,11 @@ def run(background=true)
             @reactor.add client
           end
         end
+
+        process_now
```
Comment (wjordan) [inline lib/puma/server.rb:228]: This return value of the `ThreadPool` block (`process_now` is true when a complete request was processed by the block) is used to determine if an OOB hook should be scheduled to fire once the thread is eventually idle.

File: lib/puma/server.rb (hunk 1 above)
Comment (wjordan) [reply lib/puma/server.rb:333]: hmm, the original `ECONNABORTED` rescue clause had `io.close` not `sock.close`, I should change this back to the original behavior.

That  said, the line is _ancient_, dating back to 2006 (https://github.com/puma/puma/commit/ab3c8082de82e6fc96838d444be06432620743ab) (!) and never covered by any tests, so it's very likely this is 'superstitious' code that should probably be changed/removed separately anyway. (In my reading of the current code, `ECONNABORTED` would only ever be thrown by `accept_nonblock` if the listener socket is closed, but in that case `io` would be `nil` causing `io.close` to always fail..)

Comment (wjordan) [reply lib/puma/server.rb]: I managed to figure out how to stub the `accept` call without needing to extract the method, so I'll move this refactoring/cleanup to a separate PR.

File: test/test_out_of_band_server.rb [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        @mutex.synchronize do
+          @oob_count += 1
+          @oob_finished.signal
+          @oob_finished.wait(@mutex, 1) if oob_wait
+        end
+      end
+    end
+    app_wait = options.delete(:app_wait)
+    app = ->(_) do
+      raise 'OOB conflict' if in_oob.locked?
```
Comment (nateberkopec) [inline test/test_out_of_band_server.rb:53]: Nice, I like that approach.

**Comment by nateberkopec** [comment]
Tests are 😍
